In [2]:
import pandas as pd
import numpy as np
import zipfile
from scipy import stats

In [3]:
df_desc = pd.read_csv('content/data_dictionary.csv')
df_desc

Instrument  \
0                           Identifier   
1                         Demographics   
2                         Demographics   
3                         Demographics   
4   Children's Global Assessment Scale   
..                                 ...   
76             Sleep Disturbance Scale   
77             Sleep Disturbance Scale   
78             Sleep Disturbance Scale   
79                        Internet Use   
80                        Internet Use   

                                     Field                       Description  \
0                                       id                  Participant's ID   
1                Basic_Demos-Enroll_Season              Season of enrollment   
2                          Basic_Demos-Age                Age of participant   
3                          Basic_Demos-Sex                Sex of participant   
4                              CGAS-Season           Season of participation   
..                                     ...                               ...   
76                              SDS-Season           Season of participation   
77                       SDS-SDS_Total_Raw                   Total Raw Score   
78                         SDS-SDS_Total_T                     Total T-Score   
79                     PreInt_EduHx-Season           Season of participation   
80  PreInt_EduHx-computerinternet_hoursday  Hours of using computer/internet   

               Type                        Values  \
0               str                           NaN   
1               str  Spring, Summer, Fall, Winter   
2             float                           NaN   
3   categorical int                           0,1   
4               str  Spring, Summer, Fall, Winter   
..              ...                           ...   
76              str  Spring, Summer, Fall, Winter   
77              int                           NaN   
78              int                           NaN   
79              str  Spring, Summer, Fall, Winter   
80  categorical int                       0,1,2,3   

                                         Value Labels  
0                                                 NaN  
1                                                 NaN  
2                                                 NaN  
3                                    0=Male, 1=Female  
4                                                 NaN  
..                                                ...  
76                                                NaN  
77                                                NaN  
78                                                NaN  
79                                                NaN  
80  0=Less than 1h/day, 1=Around 1h/day, 2=Around ...  

[81 rows x 6 columns]

In [4]:
df = pd.read_csv('content/train.csv')
df

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0     00008ff9                      Fall                5                0   
1     000fd460                    Summer                9                0   
2     00105258                    Summer               10                1   
3     00115b9f                    Winter                9                0   
4     0016bb22                    Spring               18                1   
...        ...                       ...              ...              ...   
3955  ff8a2de4                      Fall               13                0   
3956  ffa9794a                    Winter               10                0   
3957  ffcd4dbd                      Fall               11                0   
3958  ffed1dd5                    Spring               13                0   
3959  ffef538e                    Spring               11                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
0         Winter             51.0            Fall     16.877316   
1            NaN              NaN            Fall     14.035590   
2           Fall             71.0            Fall     16.648696   
3           Fall             71.0          Summer     18.292347   
4         Summer              NaN             NaN           NaN   
...          ...              ...             ...           ...   
3955      Spring             60.0            Fall     16.362460   
3956         NaN              NaN          Spring     18.764678   
3957      Spring             68.0          Winter     21.441500   
3958      Spring             70.0          Winter     12.235895   
3959         NaN              NaN          Winter           NaN   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
0                46.0             50.8  ...             4.0             2.0   
1                48.0             46.0  ...             0.0             0.0   
2                56.5             75.6  ...             2.0             1.0   
3                56.0             81.6  ...             3.0             4.0   
4                 NaN              NaN  ...             NaN             NaN   
...               ...              ...  ...             ...             ...   
3955             59.5             82.4  ...             1.0             1.0   
3956             53.5             76.4  ...             NaN             NaN   
3957             60.0            109.8  ...             1.0             0.0   
3958             70.7             87.0  ...             1.0             1.0   
3959              NaN              NaN  ...             NaN             NaN   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season  SDS-SDS_Total_Raw  \
0                4.0               55.0        NaN                NaN   
1                0.0                0.0       Fall               46.0   
2                1.0               28.0       Fall               38.0   
3                1.0               44.0     Summer               31.0   
4                NaN                NaN        NaN                NaN   
...              ...                ...        ...                ...   
3955             0.0               32.0     Winter               35.0   
3956             NaN                NaN        NaN                NaN   
3957             1.0               31.0     Winter               56.0   
3958             1.0               19.0     Spring               33.0   
3959             NaN                NaN        NaN                NaN   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
0                 NaN                 Fall   
1                64.0               Summer   
2                54.0               Summer   
3                45.0               Winter   
4                 NaN                  NaN   
...               ...                  ...   
3955             50.0                 Fall   
3956              NaN               Winter   
3957             77.0         

In [5]:
df.shape

(3960, 82)

# Before we begin...

Our data has categorical information, true numerical information, and categorical information stored as numbers

In [6]:
df_desc.head() # look at type column

Instrument                      Field  \
0                          Identifier                         id   
1                        Demographics  Basic_Demos-Enroll_Season   
2                        Demographics            Basic_Demos-Age   
3                        Demographics            Basic_Demos-Sex   
4  Children's Global Assessment Scale                CGAS-Season   

               Description             Type                        Values  \
0         Participant's ID              str                           NaN   
1     Season of enrollment              str  Spring, Summer, Fall, Winter   
2       Age of participant            float                           NaN   
3       Sex of participant  categorical int                           0,1   
4  Season of participation              str  Spring, Summer, Fall, Winter   

       Value Labels  
0               NaN  
1               NaN  
2               NaN  
3  0=Male, 1=Female  
4               NaN

In [7]:
df_desc['Type'].unique()

array(['str', 'float', 'categorical int', 'int'], dtype=object)

In [8]:
df_desc[df_desc['Field'] == 'sii'] # worthy to note that the description table doesn't have a row for the output sii

Empty DataFrame
Columns: [Instrument, Field, Description, Type, Values, Value Labels]
Index: []

In [9]:
# categoricals = df.dtypes[df.dtypes == 'object'].index
categoricals = pd.Index(df_desc[df_desc['Type'] == 'str']['Field'])
categoricals # seasonal data + id

Index(['id', 'Basic_Demos-Enroll_Season', 'CGAS-Season', 'Physical-Season',
       'Fitness_Endurance-Season', 'FGC-Season', 'BIA-Season', 'PAQ_A-Season',
       'PAQ_C-Season', 'PCIAT-Season', 'SDS-Season', 'PreInt_EduHx-Season'],
      dtype='object', name='Field')

In [10]:
# allnumericals = df.dtypes[df.dtypes != 'object'].index
allnumericals = pd.Index(df_desc[df_desc['Type'] != 'str']['Field'])
allnumericals

Index(['Basic_Demos-Age', 'Basic_Demos-Sex', 'CGAS-CGAS_Score', 'Physical-BMI',
       'Physical-Height', 'Physical-Weight', 'Physical-Waist_Circumference',
       'Physical-Diastolic_BP', 'Physical-HeartRate', 'Physical-Systolic_BP',
       'Fitness_Endurance-Max_Stage', 'Fitness_Endurance-Time_Mins',
       'Fitness_Endurance-Time_Sec', 'FGC-FGC_CU', 'FGC-FGC_CU_Zone',
       'FGC-FGC_GSND', 'FGC-FGC_GSND_Zone', 'FGC-FGC_GSD', 'FGC-FGC_GSD_Zone',
       'FGC-FGC_PU', 'FGC-FGC_PU_Zone', 'FGC-FGC_SRL', 'FGC-FGC_SRL_Zone',
       'FGC-FGC_SRR', 'FGC-FGC_SRR_Zone', 'FGC-FGC_TL', 'FGC-FGC_TL_Zone',
       'BIA-BIA_Activity_Level_num', 'BIA-BIA_BMC', 'BIA-BIA_BMI',
       'BIA-BIA_BMR', 'BIA-BIA_DEE', 'BIA-BIA_ECW', 'BIA-BIA_FFM',
       'BIA-BIA_FFMI', 'BIA-BIA_FMI', 'BIA-BIA_Fat', 'BIA-BIA_Frame_num',
       'BIA-BIA_ICW', 'BIA-BIA_LDM', 'BIA-BIA_LST', 'BIA-BIA_SMM',
       'BIA-BIA_TBW', 'PAQ_A-PAQ_A_Total', 'PAQ_C-PAQ_C_Total',
       'PCIAT-PCIAT_01', 'PCIAT-PCIAT_02', 'PCIAT-PCIAT_03

In [11]:
# numcategoricals = allnumericals[(df[allnumericals].min() == 0) & (df[allnumericals].max() <= 5)]
numcategoricals = pd.Index(df_desc[df_desc['Type'] == 'categorical int']['Field']).append(pd.Index(['sii'],name='Field'))
numcategoricals

Index(['Basic_Demos-Sex', 'FGC-FGC_CU_Zone', 'FGC-FGC_GSND_Zone',
       'FGC-FGC_GSD_Zone', 'FGC-FGC_PU_Zone', 'FGC-FGC_SRL_Zone',
       'FGC-FGC_SRR_Zone', 'FGC-FGC_TL_Zone', 'BIA-BIA_Activity_Level_num',
       'BIA-BIA_Frame_num', 'PCIAT-PCIAT_01', 'PCIAT-PCIAT_02',
       'PCIAT-PCIAT_03', 'PCIAT-PCIAT_04', 'PCIAT-PCIAT_05', 'PCIAT-PCIAT_06',
       'PCIAT-PCIAT_07', 'PCIAT-PCIAT_08', 'PCIAT-PCIAT_09', 'PCIAT-PCIAT_10',
       'PCIAT-PCIAT_11', 'PCIAT-PCIAT_12', 'PCIAT-PCIAT_13', 'PCIAT-PCIAT_14',
       'PCIAT-PCIAT_15', 'PCIAT-PCIAT_16', 'PCIAT-PCIAT_17', 'PCIAT-PCIAT_18',
       'PCIAT-PCIAT_19', 'PCIAT-PCIAT_20',
       'PreInt_EduHx-computerinternet_hoursday', 'sii'],
      dtype='object', name='Field')

In [12]:
# numericals = allnumericals[~((df[allnumericals].min() == 0) & (df[allnumericals].max() <= 5))]
numericals = pd.Index(df_desc[(df_desc['Type'] == 'float') | (df_desc['Type'] == 'int')]['Field'])
numericals

Index(['Basic_Demos-Age', 'CGAS-CGAS_Score', 'Physical-BMI', 'Physical-Height',
       'Physical-Weight', 'Physical-Waist_Circumference',
       'Physical-Diastolic_BP', 'Physical-HeartRate', 'Physical-Systolic_BP',
       'Fitness_Endurance-Max_Stage', 'Fitness_Endurance-Time_Mins',
       'Fitness_Endurance-Time_Sec', 'FGC-FGC_CU', 'FGC-FGC_GSND',
       'FGC-FGC_GSD', 'FGC-FGC_PU', 'FGC-FGC_SRL', 'FGC-FGC_SRR', 'FGC-FGC_TL',
       'BIA-BIA_BMC', 'BIA-BIA_BMI', 'BIA-BIA_BMR', 'BIA-BIA_DEE',
       'BIA-BIA_ECW', 'BIA-BIA_FFM', 'BIA-BIA_FFMI', 'BIA-BIA_FMI',
       'BIA-BIA_Fat', 'BIA-BIA_ICW', 'BIA-BIA_LDM', 'BIA-BIA_LST',
       'BIA-BIA_SMM', 'BIA-BIA_TBW', 'PAQ_A-PAQ_A_Total', 'PAQ_C-PAQ_C_Total',
       'PCIAT-PCIAT_Total', 'SDS-SDS_Total_Raw', 'SDS-SDS_Total_T'],
      dtype='object', name='Field')

In [13]:
len(categoricals) + len(numcategoricals) + len(numericals), df.shape[1]

(82, 82)

# Cleaning up the Zeros

Some of the data is filled in with 0 when it really should be `NaN`

In [14]:
df.describe().loc['min'][df.describe().loc['min'] < 0].index

Index(['BIA-BIA_BMC', 'BIA-BIA_FMI', 'BIA-BIA_Fat'], dtype='object')

These columns have a negative minimum. Is this possible?

In [15]:
df_desc[df_desc['Field'].isin(df.describe().loc['min'][df.describe().loc['min'] < 0].index)]

Instrument        Field           Description   Type  \
35  Bio-electric Impedance Analysis  BIA-BIA_BMC  Bone Mineral Content  float   
42  Bio-electric Impedance Analysis  BIA-BIA_FMI        Fat Mass Index  float   
43  Bio-electric Impedance Analysis  BIA-BIA_Fat   Body Fat Percentage  float   

   Values Value Labels  
35    NaN          NaN  
42    NaN          NaN  
43    NaN          NaN

You cannot have negative bone mineral content, negative fat mass index, or negative body fat percentage. We should replace the negative values with `NaN`.

In [16]:
replaceNegatives = df.describe().loc['min'][df.describe().loc['min'] < 0].index
replaceNegatives

Index(['BIA-BIA_BMC', 'BIA-BIA_FMI', 'BIA-BIA_Fat'], dtype='object')

In [17]:
categoricals.intersection(df.describe().loc['min'][df.describe().loc['min'] == 0].index)

Index([], dtype='object')

In [18]:
numericals.intersection(df.describe().loc['min'][df.describe().loc['min'] == 0].index)

Index(['Physical-BMI', 'Physical-Weight', 'Physical-Diastolic_BP',
       'Physical-Systolic_BP', 'Fitness_Endurance-Max_Stage',
       'Fitness_Endurance-Time_Mins', 'Fitness_Endurance-Time_Sec',
       'FGC-FGC_CU', 'FGC-FGC_GSND', 'FGC-FGC_GSD', 'FGC-FGC_PU',
       'FGC-FGC_SRL', 'FGC-FGC_SRR', 'FGC-FGC_TL', 'PCIAT-PCIAT_Total'],
      dtype='object')

These numerical columns have a minimum of 0. Is this possible?  
For most of them, they should not have values of 0. For example `Physical-Weight` of 0 does not mean anything, so we should replace 0s with `NaN`.  
However, the one exception is the column `PCIAT-PCIAT_Total`, which is the total score to the Parent-Child Internet Addiction Test, where it is possible to have a score of 0.

In [19]:
numcategoricals.intersection(df.describe().loc['min'][df.describe().loc['min'] == 0].index)

Index(['Basic_Demos-Sex', 'FGC-FGC_CU_Zone', 'FGC-FGC_PU_Zone',
       'FGC-FGC_SRL_Zone', 'FGC-FGC_SRR_Zone', 'FGC-FGC_TL_Zone',
       'PCIAT-PCIAT_01', 'PCIAT-PCIAT_02', 'PCIAT-PCIAT_03', 'PCIAT-PCIAT_04',
       'PCIAT-PCIAT_05', 'PCIAT-PCIAT_06', 'PCIAT-PCIAT_07', 'PCIAT-PCIAT_08',
       'PCIAT-PCIAT_09', 'PCIAT-PCIAT_10', 'PCIAT-PCIAT_11', 'PCIAT-PCIAT_12',
       'PCIAT-PCIAT_13', 'PCIAT-PCIAT_14', 'PCIAT-PCIAT_15', 'PCIAT-PCIAT_16',
       'PCIAT-PCIAT_17', 'PCIAT-PCIAT_18', 'PCIAT-PCIAT_19', 'PCIAT-PCIAT_20',
       'PreInt_EduHx-computerinternet_hoursday', 'sii'],
      dtype='object')

These categorical columns have a minimum of 0. Is this possible?  
Most likely yes, since even though they are recorded as numbers, they are 'categorical'. For example, `Basic_Demos-Sex` is recorded as 0 and 1.

In [20]:
replaceZeros = numericals.intersection(df.describe().loc['min'][df.describe().loc['min'] == 0].index).drop('PCIAT-PCIAT_Total')
replaceZeros

Index(['Physical-BMI', 'Physical-Weight', 'Physical-Diastolic_BP',
       'Physical-Systolic_BP', 'Fitness_Endurance-Max_Stage',
       'Fitness_Endurance-Time_Mins', 'Fitness_Endurance-Time_Sec',
       'FGC-FGC_CU', 'FGC-FGC_GSND', 'FGC-FGC_GSD', 'FGC-FGC_PU',
       'FGC-FGC_SRL', 'FGC-FGC_SRR', 'FGC-FGC_TL'],
      dtype='object')

In [21]:
df_desc[df_desc['Field'].isin(replaceZeros)]

Instrument                        Field  \
7                  Physical Measures                 Physical-BMI   
9                  Physical Measures              Physical-Weight   
11                 Physical Measures        Physical-Diastolic_BP   
13                 Physical Measures         Physical-Systolic_BP   
15  FitnessGram Vitals and Treadmill  Fitness_Endurance-Max_Stage   
16  FitnessGram Vitals and Treadmill  Fitness_Endurance-Time_Mins   
17  FitnessGram Vitals and Treadmill   Fitness_Endurance-Time_Sec   
19                 FitnessGram Child                   FGC-FGC_CU   
21                 FitnessGram Child                 FGC-FGC_GSND   
23                 FitnessGram Child                  FGC-FGC_GSD   
25                 FitnessGram Child                   FGC-FGC_PU   
27                 FitnessGram Child                  FGC-FGC_SRL   
29                 FitnessGram Child                  FGC-FGC_SRR   
31                 FitnessGram Child                   FGC-FGC_TL   

                           Description   Type Values Value Labels  
7             Body Mass Index (kg/m^2)  float    NaN          NaN  
9                         Weight (lbs)  float    NaN          NaN  
11                 Diastolic BP (mmHg)    int    NaN          NaN  
13                  Systolic BP (mmHg)    int    NaN          NaN  
15               Maximum stage reached    int    NaN          NaN  
16       Exact time completed: Minutes    int    NaN          NaN  
17       Exact time completed: Seconds    int    NaN          NaN  
19                       Curl up total    int    NaN          NaN  
21  Grip Strength total (non-dominant)  float    NaN          NaN  
23      Grip Strength total (dominant)  float    NaN          NaN  
25                       Push-up total    int    NaN          NaN  
27       Sit & Reach total (left side)  float    NaN          NaN  
29      Sit & Reach total (right side)  float    NaN          NaN  
31                    Trunk lift total    int    NaN          NaN

A value of 0 is impossible for BMI, weight, blood pressure, or any fitness test that counts reps.

We should replace these 0 values with `NaN`.

In [22]:
df[replaceNegatives] = df[replaceNegatives].mask(df[replaceNegatives] <= 0)

In [23]:
(df[replaceNegatives] <= 0).sum().sum()

0

In [24]:
df[replaceZeros] = df[replaceZeros].mask(df[replaceZeros] <= 0)

In [25]:
(df[replaceZeros] <= 0).sum().sum()

0

In [26]:
df.isnull().sum().sum() / (df.shape[0] * df.shape[1])

0.4108770633160877

41% of the data is now null

In [27]:
# df[(df[df.dtypes.index[df.dtypes != 'object']] < 0).any(axis=1)]

In [28]:
# df[df.dtypes.index[df.dtypes != 'object']] < 0

In [29]:
# df.mask(df[df.dtypes.index[df.dtypes != 'object']] < 0, np.nan)
# df[df.dtypes.index[df.dtypes != 'object']].mask(df[df.dtypes.index[df.dtypes != 'object']] < 0, np.nan)

## Removing columns in which more than 80% of the rows are null

In [30]:
threshold = 0.8

In [31]:
(df.isnull().sum() > len(df)*threshold).sum()

5

There are 5 columns who have >80% missing values.

In [32]:
removeCols = df.columns[df.isnull().sum() > len(df)*threshold]

In [33]:
df_desc[df_desc['Field'].isin(removeCols)]

Instrument  \
15               FitnessGram Vitals and Treadmill   
16               FitnessGram Vitals and Treadmill   
17               FitnessGram Vitals and Treadmill   
50  Physical Activity Questionnaire (Adolescents)   
51  Physical Activity Questionnaire (Adolescents)   

                          Field                           Description   Type  \
15  Fitness_Endurance-Max_Stage                 Maximum stage reached    int   
16  Fitness_Endurance-Time_Mins         Exact time completed: Minutes    int   
17   Fitness_Endurance-Time_Sec         Exact time completed: Seconds    int   
50                 PAQ_A-Season               Season of participation    str   
51            PAQ_A-PAQ_A_Total  Activity Summary Score (Adolescents)  float   

                          Values Value Labels  
15                           NaN          NaN  
16                           NaN          NaN  
17                           NaN          NaN  
50  Spring, Summer, Fall, Winter          NaN  
51                           NaN          NaN

In [34]:
"""
filter_null = df.isnull().sum().sort_values(ascending = True) <= (len(df) / 2)
filter_null = filter_null[filter_null == True]
list_of_columns_less_null = filter_null.index.to_list()
df_filter_null = df[list_of_columns_less_null]
df_filter_null.isna().sum().sum() / len(df_filter_null)
len(df_filter_null)
"""
df = df.drop(columns=removeCols)
df.head()

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0  00008ff9                      Fall                5                0   
1  000fd460                    Summer                9                0   
2  00105258                    Summer               10                1   
3  00115b9f                    Winter                9                0   
4  0016bb22                    Spring               18                1   

  CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  Physical-Height  \
0      Winter             51.0            Fall     16.877316             46.0   
1         NaN              NaN            Fall     14.035590             48.0   
2        Fall             71.0            Fall     16.648696             56.5   
3        Fall             71.0          Summer     18.292347             56.0   
4      Summer              NaN             NaN           NaN              NaN   

   Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  PCIAT-PCIAT_20  \
0             50.8  ...             4.0             2.0             4.0   
1             46.0  ...             0.0             0.0             0.0   
2             75.6  ...             2.0             1.0             1.0   
3             81.6  ...             3.0             4.0             1.0   
4              NaN  ...             NaN             NaN             NaN   

   PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  SDS-SDS_Total_T  \
0               55.0        NaN               NaN              NaN   
1                0.0       Fall              46.0             64.0   
2               28.0       Fall              38.0             54.0   
3               44.0     Summer              31.0             45.0   
4                NaN        NaN               NaN              NaN   

   PreInt_EduHx-Season  PreInt_EduHx-computerinternet_hoursday  sii  
0                 Fall                                     3.0  2.0  
1               Summer                                     0.0  0.0  
2               Summer                                     2.0  0.0  
3               Winter                                     0.0  1.0  
4                  NaN                                     NaN  NaN  

[5 rows x 77 columns]

In [35]:
df.isnull().sum().sum() / (df.shape[0] * df.shape[1])

0.38279876688967596

38% of the data is now null.

In [36]:
(df.isnull().sum(axis=1) > df.shape[1]*threshold).sum()

756

756 rows of the data have more than 80% of the attributes missing! Consider removing those rows.  
On the other hand, we also have time series data as well for some of the rows, so we can wait and see later on if we really need to remove the rows that have a lot of missing values.

In [37]:
#numericals = numericals.drop(['Fitness_Endurance-Max_Stage','Fitness_Endurance-Time_Mins','Fitness_Endurance-Time_Sec','PAQ_A-PAQ_A_Total'])

In [38]:
#categoricals = categoricals.drop(['PAQ_A-Season'])

In [39]:
df_desc = df_desc[~(df_desc['Field'].isin(removeCols))]

In [40]:
numericals = pd.Index(df_desc[(df_desc['Type'] == 'float') | (df_desc['Type'] == 'int')]['Field'])
categoricals = pd.Index(df_desc[df_desc['Type'] == 'str']['Field'])
numcategoricals = pd.Index(df_desc[df_desc['Type'] == 'categorical int']['Field']).append(pd.Index(['sii'],name='Field'))

# Removing the outliers

In [41]:
df.describe()

Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  Physical-BMI  \
count      3960.000000      3960.000000      2421.000000   3015.000000   
mean         10.433586         0.372727        65.454771     19.376812   
std           3.574648         0.483591        22.341862      5.034191   
min           5.000000         0.000000        25.000000      8.522436   
25%           8.000000         0.000000        59.000000     15.890526   
50%          10.000000         0.000000        65.000000     17.950925   
75%          13.000000         1.000000        75.000000     21.588631   
max          22.000000         1.000000       999.000000     59.132048   

       Physical-Height  Physical-Weight  Physical-Waist_Circumference  \
count      3027.000000      3015.000000                    898.000000   
mean         55.946713        90.840060                     27.278508   
std           7.473764        43.161374                      5.567287   
min          33.000000        31.800000                     18.000000   
25%          50.000000        58.200000                     23.000000   
50%          55.000000        77.800000                     26.000000   
75%          62.000000       114.300000                     30.000000   
max          78.500000       315.000000                     50.000000   

       Physical-Diastolic_BP  Physical-HeartRate  Physical-Systolic_BP  ...  \
count            2953.000000         2967.000000           2953.000000  ...   
mean               69.672536           81.597236            117.022689  ...   
std                13.553022           13.665196             16.927687  ...   
min                11.000000           27.000000             49.000000  ...   
25%                61.000000           72.000000            107.000000  ...   
50%                68.000000           81.000000            114.000000  ...   
75%                76.000000           90.500000            125.000000  ...   
max               179.000000          138.000000            203.000000  ...   

       PCIAT-PCIAT_16  PCIAT-PCIAT_17  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
count     2728.000000     2725.000000     2728.000000     2730.000000   
mean         1.452346        1.627890        1.613636        1.158974   
std          1.495600        1.445622        1.529178        1.343661   
min          0.000000        0.000000        0.000000        0.000000   
25%          0.000000        0.000000        0.000000        0.000000   
50%          1.000000        1.000000        1.000000        1.000000   
75%          2.000000        3.000000        2.000000        2.000000   
max          5.000000        5.000000        5.000000        5.000000   

       PCIAT-PCIAT_20  PCIAT-PCIAT_Total  SDS-SDS_Total_Raw  SDS-SDS_Total_T  \
count     2733.000000        2736.000000        2609.000000      2606.000000   
mean         0.943652          27.896199          41.088923        57.763622   
std          1.185460          20.338853          10.427433        13.196091   
min          0.000000           0.000000          17.000000        38.000000   
25%          0.000000          12.000000          33.000000        47.000000   
50%          1.000000          26.000000          39.000000        55.000000   
75%          1.000000          41.000000          46.000000        64.000000   
max          5.000000          93.000000          96.000000       100.000000   

       PreInt_EduHx-computerinternet_hoursday          sii  
count                             3301.000000  2736.000000  
mean                                 1.060588     0.580409  
std                                  1.094875     0.771122  
min                                  0.000000     0.000000  
25%                                  0.000000     0.000000  
50%                                  1.000000     0.000000  
75%                                  2.000000     1.000000  
max                                  3.000000     3.000000  

[8 rows x 66 columns]

In [42]:
df[(np.abs(stats.zscore(df[numericals], nan_policy='omit')) >= 3).any(axis=1)]

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
6     0038ba98                      Fall               10                0   
51    02cebf33                    Spring               12                1   
59    0380e709                    Summer               17                0   
87    04df895a                    Winter               17                0   
193   0acac70c                    Spring               18                0   
...        ...                       ...              ...              ...   
3876  f9ca9ff6                    Winter               11                1   
3903  fb9b060f                    Summer               16                1   
3922  fcd9a6bb                    Spring                6                0   
3946  feb53635                    Winter               14                0   
3952  ff60112d                    Summer               15                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
6            NaN              NaN            Fall     19.660760   
51        Spring             77.0          Summer     19.751301   
59           NaN              NaN          Summer     24.446886   
87           NaN              NaN          Winter     26.630764   
193         Fall             65.0          Summer     21.485616   
...          ...              ...             ...           ...   
3876         NaN              NaN          Spring     16.319643   
3903        Fall             65.0          Summer     21.266964   
3922      Summer             69.0          Spring     16.195145   
3946      Summer             70.0          Spring     19.036779   
3952      Spring             40.0          Winter     26.364710   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
6               55.00             84.6  ...             4.0             1.0   
51              62.00            108.0  ...             3.0             3.0   
59              68.00            160.8  ...             4.0             2.0   
87              77.00            224.6  ...             NaN             NaN   
193             70.15            150.4  ...             0.0             3.0   
...               ...              ...  ...             ...             ...   
3876            56.00             72.8  ...             NaN             NaN   
3903            67.00            135.8  ...             3.0             2.0   
3922            44.00             44.6  ...             1.0             0.0   
3946            69.50            130.8  ...             4.0             4.0   
3952            70.50            186.4  ...             1.0             1.0   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
6                0.0               20.0     Winter              27.0   
51               3.0               46.0     Summer              82.0   
59               2.0               56.0        NaN               NaN   
87               NaN                NaN        NaN               NaN   
193              4.0               34.0     Summer              33.0   
...              ...                ...        ...               ...   
3876             NaN                NaN        NaN               NaN   
3903             1.0               52.0     Summer              68.0   
3922             0.0               14.0     Spring              38.0   
3946             4.0               77.0     Spring              51.0   
3952             1.0               35.0     Spring              66.0   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
6                40.0                 Fall   
51              100.0               Spring   
59                NaN               Summer   
87                NaN               Winter   
193              47.0               Spring   
...               ...                  ...   
3876              NaN               Winter   
3903             93.0               Summer   
3922             54.0               Spring

In [43]:
# df[((np.abs(stats.zscore(df[numericals], nan_policy='omit')) < 3) | (np.abs(stats.zscore(df[numericals], nan_policy='omit')).isnull())).all(axis=1)]

If we track outliers based on a $|zscore| >= 3$ metric, then that takes out 317 rows of the data!

In [44]:
317 / df.shape[0]

0.08005050505050505

8% outliers based on zscore

But, the data is mostly skewed, so perhaps we can try something other than zscore for outlier detection.  
How about based on IQR?

In [45]:
Q1 = df[numericals].quantile(0.25)
Q3 = df[numericals].quantile(0.75)
IQR = Q3 - Q1

In [46]:
df[((df[numericals] < (Q1 - 1.5 * IQR)) | (df[numericals] > (Q3 + 1.5 * IQR))).any(axis=1)]

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
5     001f3379                    Spring               13                1   
6     0038ba98                      Fall               10                0   
21    00f574e9                    Summer               16                0   
22    01085eb3                      Fall               12                0   
32    01d24f60                      Fall               16                0   
...        ...                       ...              ...              ...   
3935  fdd9799f                    Spring               15                0   
3938  fe102a25                    Summer                5                0   
3945  fe9c71d8                    Winter                9                0   
3946  feb53635                    Winter               14                0   
3952  ff60112d                    Summer               15                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
5         Winter             50.0          Summer     22.279952   
6            NaN              NaN            Fall     19.660760   
21           NaN              NaN          Summer     29.285856   
22        Winter             58.0            Fall     34.187282   
32           NaN              NaN             NaN           NaN   
...          ...              ...             ...           ...   
3935      Summer             48.0          Summer     29.404637   
3938      Winter             61.0          Summer     18.278557   
3945      Spring             80.0          Winter     15.891137   
3946      Summer             70.0          Spring     19.036779   
3952      Spring             40.0          Winter     26.364710   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
5               59.50            112.2  ...             1.0             2.0   
6               55.00             84.6  ...             4.0             1.0   
21              71.00            210.0  ...             2.0             1.0   
22              60.50            178.0  ...             1.0             1.0   
32                NaN              NaN  ...             NaN             NaN   
...               ...              ...  ...             ...             ...   
3935            66.00            182.2  ...             2.0             4.0   
3938            43.50             49.2  ...             3.0             0.0   
3945            56.75             72.8  ...             1.0             1.0   
3946            69.50            130.8  ...             4.0             4.0   
3952            70.50            186.4  ...             1.0             1.0   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
5                1.0               34.0     Summer              40.0   
6                0.0               20.0     Winter              27.0   
21               0.0               30.0        NaN               NaN   
22               0.0               30.0       Fall              53.0   
32               NaN                NaN     Winter              67.0   
...              ...                ...        ...               ...   
3935             2.0               45.0     Summer              40.0   
3938             0.0               34.0        NaN               NaN   
3945             1.0               29.0     Winter              33.0   
3946             4.0               77.0     Spring              51.0   
3952             1.0               35.0     Spring              66.0   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
5                56.0               Spring   
6                40.0                 Fall   
21                NaN               Summer   
22               73.0                 Fall   
32               91.0                 Fall   
...               ...                  ...   
3935             56.0               Spring   
3938              NaN               Summer   
3945             47.0               Winter

In [47]:
784 / df.shape[0]

0.19797979797979798

20% outliers according to using the IQR!  
That's a lot of the data considered as outliers, perhaps we should widen the outlier fence boundaries

In [48]:
df[((df[numericals] < (Q1 - 2.5 * IQR)) | (df[numericals] > (Q3 + 2.5 * IQR))).any(axis=1)]

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
6     0038ba98                      Fall               10                0   
22    01085eb3                      Fall               12                0   
51    02cebf33                    Spring               12                1   
59    0380e709                    Summer               17                0   
144   0844aee7                    Summer                8                0   
...        ...                       ...              ...              ...   
3844  f7fc97cf                    Spring               14                0   
3863  f8ff0bc8                      Fall               18                0   
3922  fcd9a6bb                    Spring                6                0   
3946  feb53635                    Winter               14                0   
3952  ff60112d                    Summer               15                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
6            NaN              NaN            Fall     19.660760   
22        Winter             58.0            Fall     34.187282   
51        Spring             77.0          Summer     19.751301   
59           NaN              NaN          Summer     24.446886   
144         Fall             61.0            Fall     14.706858   
...          ...              ...             ...           ...   
3844      Spring             75.0          Spring     23.752311   
3863      Spring             62.0          Winter     26.429967   
3922      Summer             69.0          Spring     16.195145   
3946      Summer             70.0          Spring     19.036779   
3952      Spring             40.0          Winter     26.364710   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
6                55.0             84.6  ...             4.0             1.0   
22               60.5            178.0  ...             1.0             1.0   
51               62.0            108.0  ...             3.0             3.0   
59               68.0            160.8  ...             4.0             2.0   
144              48.0             48.2  ...             1.0             1.0   
...               ...              ...  ...             ...             ...   
3844             69.5            163.2  ...             1.0             1.0   
3863             71.5            192.2  ...             1.0             1.0   
3922             44.0             44.6  ...             1.0             0.0   
3946             69.5            130.8  ...             4.0             4.0   
3952             70.5            186.4  ...             1.0             1.0   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
6                0.0               20.0     Winter              27.0   
22               0.0               30.0       Fall              53.0   
51               3.0               46.0     Summer              82.0   
59               2.0               56.0        NaN               NaN   
144              1.0               32.0       Fall              32.0   
...              ...                ...        ...               ...   
3844             2.0               23.0     Spring              35.0   
3863             2.0               24.0     Winter              42.0   
3922             0.0               14.0     Spring              38.0   
3946             4.0               77.0     Spring              51.0   
3952             1.0               35.0     Spring              66.0   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
6                40.0                 Fall   
22               73.0                 Fall   
51              100.0               Spring   
59                NaN               Summer   
144              46.0               Summer   
...               ...                  ...   
3844             50.0               Spring   
3863             59.0                 Fall   
3922             54.0               Spring

In [49]:
274 / df.shape[0]

0.0691919191919192

Let's remove this 7% of the data that is extreme

In [50]:
df = df[~(((df[numericals] < (Q1 - 1.5 * IQR)) | (df[numericals] > (Q3 + 1.5 * IQR))).any(axis=1))]
df

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0     00008ff9                      Fall                5                0   
1     000fd460                    Summer                9                0   
2     00105258                    Summer               10                1   
3     00115b9f                    Winter                9                0   
4     0016bb22                    Spring               18                1   
...        ...                       ...              ...              ...   
3955  ff8a2de4                      Fall               13                0   
3956  ffa9794a                    Winter               10                0   
3957  ffcd4dbd                      Fall               11                0   
3958  ffed1dd5                    Spring               13                0   
3959  ffef538e                    Spring               11                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
0         Winter             51.0            Fall     16.877316   
1            NaN              NaN            Fall     14.035590   
2           Fall             71.0            Fall     16.648696   
3           Fall             71.0          Summer     18.292347   
4         Summer              NaN             NaN           NaN   
...          ...              ...             ...           ...   
3955      Spring             60.0            Fall     16.362460   
3956         NaN              NaN          Spring     18.764678   
3957      Spring             68.0          Winter     21.441500   
3958      Spring             70.0          Winter     12.235895   
3959         NaN              NaN          Winter           NaN   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
0                46.0             50.8  ...             4.0             2.0   
1                48.0             46.0  ...             0.0             0.0   
2                56.5             75.6  ...             2.0             1.0   
3                56.0             81.6  ...             3.0             4.0   
4                 NaN              NaN  ...             NaN             NaN   
...               ...              ...  ...             ...             ...   
3955             59.5             82.4  ...             1.0             1.0   
3956             53.5             76.4  ...             NaN             NaN   
3957             60.0            109.8  ...             1.0             0.0   
3958             70.7             87.0  ...             1.0             1.0   
3959              NaN              NaN  ...             NaN             NaN   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
0                4.0               55.0        NaN               NaN   
1                0.0                0.0       Fall              46.0   
2                1.0               28.0       Fall              38.0   
3                1.0               44.0     Summer              31.0   
4                NaN                NaN        NaN               NaN   
...              ...                ...        ...               ...   
3955             0.0               32.0     Winter              35.0   
3956             NaN                NaN        NaN               NaN   
3957             1.0               31.0     Winter              56.0   
3958             1.0               19.0     Spring              33.0   
3959             NaN                NaN        NaN               NaN   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
0                 NaN                 Fall   
1                64.0               Summer   
2                54.0               Summer   
3                45.0               Winter   
4                 NaN                  NaN   
...               ...                  ...   
3955             50.0                 Fall   
3956              NaN               Winter   
3957             77.0                 Fall

## Impute NaN values with random values from the same column

In [51]:
df.isnull().sum().sum()

102547

Let's impute with random values from the same column

In [52]:
df

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0     00008ff9                      Fall                5                0   
1     000fd460                    Summer                9                0   
2     00105258                    Summer               10                1   
3     00115b9f                    Winter                9                0   
4     0016bb22                    Spring               18                1   
...        ...                       ...              ...              ...   
3955  ff8a2de4                      Fall               13                0   
3956  ffa9794a                    Winter               10                0   
3957  ffcd4dbd                      Fall               11                0   
3958  ffed1dd5                    Spring               13                0   
3959  ffef538e                    Spring               11                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
0         Winter             51.0            Fall     16.877316   
1            NaN              NaN            Fall     14.035590   
2           Fall             71.0            Fall     16.648696   
3           Fall             71.0          Summer     18.292347   
4         Summer              NaN             NaN           NaN   
...          ...              ...             ...           ...   
3955      Spring             60.0            Fall     16.362460   
3956         NaN              NaN          Spring     18.764678   
3957      Spring             68.0          Winter     21.441500   
3958      Spring             70.0          Winter     12.235895   
3959         NaN              NaN          Winter           NaN   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
0                46.0             50.8  ...             4.0             2.0   
1                48.0             46.0  ...             0.0             0.0   
2                56.5             75.6  ...             2.0             1.0   
3                56.0             81.6  ...             3.0             4.0   
4                 NaN              NaN  ...             NaN             NaN   
...               ...              ...  ...             ...             ...   
3955             59.5             82.4  ...             1.0             1.0   
3956             53.5             76.4  ...             NaN             NaN   
3957             60.0            109.8  ...             1.0             0.0   
3958             70.7             87.0  ...             1.0             1.0   
3959              NaN              NaN  ...             NaN             NaN   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
0                4.0               55.0        NaN               NaN   
1                0.0                0.0       Fall              46.0   
2                1.0               28.0       Fall              38.0   
3                1.0               44.0     Summer              31.0   
4                NaN                NaN        NaN               NaN   
...              ...                ...        ...               ...   
3955             0.0               32.0     Winter              35.0   
3956             NaN                NaN        NaN               NaN   
3957             1.0               31.0     Winter              56.0   
3958             1.0               19.0     Spring              33.0   
3959             NaN                NaN        NaN               NaN   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
0                 NaN                 Fall   
1                64.0               Summer   
2                54.0               Summer   
3                45.0               Winter   
4                 NaN                  NaN   
...               ...                  ...   
3955             50.0                 Fall   
3956              NaN               Winter   
3957             77.0                 Fall

In [53]:
for col in df.columns:
    choices = df[col].dropna().values
    
    nan_indices = df[col].isnull()
    df.loc[nan_indices, col] = np.random.choice(choices, size=nan_indices.sum())

In [54]:
df.isnull().sum().sum()

0

In [55]:
df

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0     00008ff9                      Fall                5                0   
1     000fd460                    Summer                9                0   
2     00105258                    Summer               10                1   
3     00115b9f                    Winter                9                0   
4     0016bb22                    Spring               18                1   
...        ...                       ...              ...              ...   
3955  ff8a2de4                      Fall               13                0   
3956  ffa9794a                    Winter               10                0   
3957  ffcd4dbd                      Fall               11                0   
3958  ffed1dd5                    Spring               13                0   
3959  ffef538e                    Spring               11                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
0         Winter             51.0            Fall     16.877316   
1         Summer             60.0            Fall     14.035590   
2           Fall             71.0            Fall     16.648696   
3           Fall             71.0          Summer     18.292347   
4         Summer             55.0          Summer     21.628730   
...          ...              ...             ...           ...   
3955      Spring             60.0            Fall     16.362460   
3956      Spring             60.0          Spring     18.764678   
3957      Spring             68.0          Winter     21.441500   
3958      Spring             70.0          Winter     12.235895   
3959      Summer             75.0          Winter     13.891280   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
0                46.0             50.8  ...             4.0             2.0   
1                48.0             46.0  ...             0.0             0.0   
2                56.5             75.6  ...             2.0             1.0   
3                56.0             81.6  ...             3.0             4.0   
4                44.0            110.2  ...             0.0             0.0   
...               ...              ...  ...             ...             ...   
3955             59.5             82.4  ...             1.0             1.0   
3956             53.5             76.4  ...             3.0             2.0   
3957             60.0            109.8  ...             1.0             0.0   
3958             70.7             87.0  ...             1.0             1.0   
3959             60.0             42.6  ...             2.0             3.0   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
0                4.0               55.0     Winter              35.0   
1                0.0                0.0       Fall              46.0   
2                1.0               28.0       Fall              38.0   
3                1.0               44.0     Summer              31.0   
4                0.0                9.0     Winter              45.0   
...              ...                ...        ...               ...   
3955             0.0               32.0     Winter              35.0   
3956             4.0               31.0       Fall              38.0   
3957             1.0               31.0     Winter              56.0   
3958             1.0               19.0     Spring              33.0   
3959             1.0                0.0     Summer              60.0   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
0                55.0                 Fall   
1                64.0               Summer   
2                54.0               Summer   
3                45.0               Winter   
4                58.0               Summer   
...               ...                  ...   
3955             50.0                 Fall   
3956             73.0               Winter   
3957             77.0                 Fall

# One Hot encoding

In [56]:
categoricals

Index(['id', 'Basic_Demos-Enroll_Season', 'CGAS-Season', 'Physical-Season',
       'Fitness_Endurance-Season', 'FGC-Season', 'BIA-Season', 'PAQ_C-Season',
       'PCIAT-Season', 'SDS-Season', 'PreInt_EduHx-Season'],
      dtype='object', name='Field')

Excluding the id, everything else is season data, so let's one hot encode them

In [57]:
categoricals = categoricals.drop('id')

In [58]:
from sklearn.preprocessing import OneHotEncoder

In [59]:
# Initialize OneHotEncoder
encoder = OneHotEncoder(sparse_output=False)
one_hot_encoded = encoder.fit_transform(df[categoricals])

one_hot_df = pd.DataFrame(one_hot_encoded,
                          columns=encoder.get_feature_names_out(categoricals))

print(one_hot_df.shape)
print(df.shape)

df.reset_index(drop=True, inplace=True)
one_hot_df.reset_index(drop=True, inplace=True)
df = pd.concat([df.drop(categoricals, axis=1), one_hot_df], axis=1)

print(df.shape)

df

(3176, 40)
(3176, 77)
(3176, 107)


id  Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  \
0     00008ff9                5                0             51.0   
1     000fd460                9                0             60.0   
2     00105258               10                1             71.0   
3     00115b9f                9                0             71.0   
4     0016bb22               18                1             55.0   
...        ...              ...              ...              ...   
3171  ff8a2de4               13                0             60.0   
3172  ffa9794a               10                0             60.0   
3173  ffcd4dbd               11                0             68.0   
3174  ffed1dd5               13                0             70.0   
3175  ffef538e               11                0             75.0   

      Physical-BMI  Physical-Height  Physical-Weight  \
0        16.877316             46.0             50.8   
1        14.035590             48.0             46.0   
2        16.648696             56.5             75.6   
3        18.292347             56.0             81.6   
4        21.628730             44.0            110.2   
...            ...              ...              ...   
3171     16.362460             59.5             82.4   
3172     18.764678             53.5             76.4   
3173     21.441500             60.0            109.8   
3174     12.235895             70.7             87.0   
3175     13.891280             60.0             42.6   

      Physical-Waist_Circumference  Physical-Diastolic_BP  Physical-HeartRate  \
0                             23.0                   44.0                82.0   
1                             22.0                   75.0                70.0   
2                             25.0                   65.0                94.0   
3                             23.0                   60.0                97.0   
4                             26.0                   74.0                83.0   
...                            ...                    ...                 ...   
3171                          24.0                   71.0                70.0   
3172                          27.0                   60.0                78.0   
3173                          30.0                   79.0                99.0   
3174                          25.0                   59.0                61.0   
3175                          22.0                   73.0                97.0   

      ...  PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
0     ...                  0.0                  0.0              0.0   
1     ...                  0.0                  0.0              1.0   
2     ...                  0.0                  0.0              1.0   
3     ...                  1.0                  0.0              0.0   
4     ...                  0.0                  0.0              0.0   
...   ...                  ...                  ...              ...   
3171  ...                  0.0                  1.0              0.0   
3172  ...                  0.0                  0.0              1.0   
3173  ...                  0.0                  1.0              0.0   
3174  ...                  0.0                  0.0              0.0   
3175  ...                  0.0                  0.0              0.0   

      SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
0                   0.0                0.0                1.0   
1                   0.0                0.0                0.0   
2                   0.0                0.0                0.0   
3                   0.0                1.0                0.0   
4                   0.0                0.0                1.0   
...                 ...                ...                ...   
3171                0.0                0.0                1.0   
3172                0.0                0.0                0.0   
3173                0.0                0.0                1.0   
3174                1.

Let's check it

In [60]:
df.iloc[:,-4:]

PreInt_EduHx-Season_Fall  PreInt_EduHx-Season_Spring  \
0                          1.0                         0.0   
1                          0.0                         0.0   
2                          0.0                         0.0   
3                          0.0                         0.0   
4                          0.0                         0.0   
...                        ...                         ...   
3171                       1.0                         0.0   
3172                       0.0                         0.0   
3173                       1.0                         0.0   
3174                       0.0                         1.0   
3175                       0.0                         1.0   

      PreInt_EduHx-Season_Summer  PreInt_EduHx-Season_Winter  
0                            0.0                         0.0  
1                            1.0                         0.0  
2                            1.0                         0.0  
3                            0.0                         1.0  
4                            1.0                         0.0  
...                          ...                         ...  
3171                         0.0                         0.0  
3172                         0.0                         1.0  
3173                         0.0                         0.0  
3174                         0.0                         0.0  
3175                         0.0                         0.0  

[3176 rows x 4 columns]

# Normalization

We'll use min max normalization since the data is skewed.

We will first split the data into train and test  
Then we'll scale the numerical columns afterward

In [61]:
from sklearn.preprocessing import MinMaxScaler

In [62]:
# shuffle the data!
df = df.sample(frac=1)
df

id  Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  \
2349  bc790ce0               12                1             35.0   
344   18350390               11                0             61.0   
1675  83912085                8                1             69.0   
1527  781598dd               20                1             50.0   
581   2a88cbe9               10                0             65.0   
...        ...              ...              ...              ...   
1943  99328d2e               12                1             60.0   
671   318ccc0e                8                0             61.0   
2671  d7c063b4                8                0             75.0   
857   4057472d               10                0             68.0   
701   333be478               11                0             70.0   

      Physical-BMI  Physical-Height  Physical-Weight  \
2349     15.055430            63.00             85.0   
344      13.286585            51.00             86.2   
1675     17.840192            54.00             74.0   
1527     26.484211            45.50             59.6   
581      20.312312            55.50             89.0   
...            ...              ...              ...   
1943     18.269680            65.00            109.8   
671      18.562044            56.00            123.4   
2671     15.952168            59.00             70.0   
857      22.432736            51.75             46.0   
701      18.477129            60.75             97.0   

      Physical-Waist_Circumference  Physical-Diastolic_BP  Physical-HeartRate  \
2349                          27.0                   64.0                83.0   
344                           31.0                   66.0                66.0   
1675                          22.0                   64.0                72.0   
1527                          23.0                   62.0                73.0   
581                           31.0                   64.0                95.0   
...                            ...                    ...                 ...   
1943                          28.0                   75.0                70.0   
671                           30.0                   66.0                65.0   
2671                          39.0                   80.0               114.0   
857                           21.0                   64.0                81.0   
701                           26.0                   66.0                72.0   

      ...  PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
2349  ...                  0.0                  0.0              1.0   
344   ...                  1.0                  0.0              0.0   
1675  ...                  0.0                  0.0              1.0   
1527  ...                  0.0                  0.0              1.0   
581   ...                  1.0                  0.0              0.0   
...   ...                  ...                  ...              ...   
1943  ...                  0.0                  0.0              1.0   
671   ...                  0.0                  0.0              1.0   
2671  ...                  1.0                  0.0              0.0   
857   ...                  0.0                  1.0              0.0   
701   ...                  0.0                  0.0              1.0   

      SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
2349                0.0                0.0                0.0   
344                 1.0                0.0                0.0   
1675                0.0                0.0                0.0   
1527                0.0                0.0                0.0   
581                 0.0                1.0                0.0   
...                 ...                ...                ...   
1943                0.0                0.0                0.0   
671                 0.0                0.0                0.0   
2671                0.0                1.0                0.0   
857                 1.

In [63]:
dfTest = df.iloc[:df.shape[0] // 10]

In [64]:
dfTrain = df.iloc[df.shape[0] // 10:]

In [65]:
dfTrain[numericals] = dfTrain[numericals].astype('float64')
dfTest[numericals] = dfTest[numericals].astype('float64')

C:\Users\kerry\AppData\Local\Temp\ipykernel_17756\2330913363.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfTrain[numericals] = dfTrain[numericals].astype('float64')
C:\Users\kerry\AppData\Local\Temp\ipykernel_17756\2330913363.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfTest[numericals] = dfTest[numericals].astype('float64')


In [66]:
scaler = MinMaxScaler()
# df_normalized = pd.DataFrame(scaler.fit_transform(df[numericals]), columns=numericals)
dfTrain.loc[:, numericals] = scaler.fit_transform(dfTrain[numericals])

dfTest.loc[:, numericals] = scaler.transform(dfTest[numericals])

In [67]:
dfTrain

id  Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  \
2868  e754411f         0.600000                1         0.250000   
1782  8c185c31         0.533333                0         0.266667   
447   1f881e3d         0.533333                0         0.250000   
141   098b5836         0.666667                0         0.750000   
918   4518ed77         0.066667                0         0.166667   
...        ...              ...              ...              ...   
1943  99328d2e         0.466667                1         0.416667   
671   318ccc0e         0.200000                0         0.433333   
2671  d7c063b4         0.200000                0         0.666667   
857   4057472d         0.333333                0         0.550000   
701   333be478         0.400000                0         0.583333   

      Physical-BMI  Physical-Height  Physical-Weight  \
2868      0.633478         0.663265         0.565657   
1782      0.455823         0.637755         0.412879   
447       0.732104         0.676020         0.652778   
141       0.626442         0.790816         0.688131   
918       0.289262         0.306122         0.566919   
...            ...              ...              ...   
1943      0.451381         0.739796         0.492424   
671       0.464920         0.510204         0.578283   
2671      0.344061         0.586735         0.241162   
857       0.644167         0.401786         0.089646   
701       0.460988         0.631378         0.411616   

      Physical-Waist_Circumference  Physical-Diastolic_BP  Physical-HeartRate  \
2868                      0.095238               0.915254            0.273973   
1782                      0.333333               0.474576            0.684932   
447                       0.904762               0.694915            0.356164   
141                       0.428571               0.372881            0.726027   
918                       0.142857               0.423729            0.383562   
...                            ...                    ...                 ...   
1943                      0.428571               0.610169            0.342466   
671                       0.523810               0.457627            0.273973   
2671                      0.952381               0.694915            0.945205   
857                       0.095238               0.423729            0.493151   
701                       0.333333               0.457627            0.369863   

      ...  PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
2868  ...                  0.0                  0.0              0.0   
1782  ...                  0.0                  0.0              0.0   
447   ...                  0.0                  0.0              0.0   
141   ...                  0.0                  1.0              0.0   
918   ...                  1.0                  0.0              0.0   
...   ...                  ...                  ...              ...   
1943  ...                  0.0                  0.0              1.0   
671   ...                  0.0                  0.0              1.0   
2671  ...                  1.0                  0.0              0.0   
857   ...                  0.0                  1.0              0.0   
701   ...                  0.0                  0.0              1.0   

      SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
2868                1.0                0.0                0.0   
1782                0.0                1.0                0.0   
447                 1.0                0.0                0.0   
141                 0.0                0.0                1.0   
918                 0.0                1.0                0.0   
...                 ...                ...                ...   
1943                0.0                0.0                0.0   
671                 0.0                0.0                0.0   
2671                0.0                1.0                0.0   
857                 1.

In [68]:
dfTest

id  Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  \
2349  bc790ce0         0.466667                1         0.000000   
344   18350390         0.400000                0         0.433333   
1675  83912085         0.200000                1         0.566667   
1527  781598dd         1.000000                1         0.250000   
581   2a88cbe9         0.333333                0         0.500000   
...        ...              ...              ...              ...   
2255  b412866b         0.533333                0         0.833333   
610   2cb0db2c         0.200000                1         0.733333   
1468  73078ded         0.133333                0         0.416667   
154   0a540797         0.066667                0         0.166667   
2964  ee74be35         0.333333                1         0.666667   

      Physical-BMI  Physical-Height  Physical-Weight  \
2349      0.302534         0.688776         0.335859   
344       0.220621         0.382653         0.343434   
1675      0.431492         0.459184         0.266414   
1527      0.831785         0.242347         0.175505   
581       0.545973         0.497449         0.361111   
...            ...              ...              ...   
2255      0.325306         0.918367         0.106061   
610       0.414396         0.076531         0.037879   
1468      0.729872         0.414541         0.230429   
154       0.488703         0.369898         0.236111   
2964      0.333788         0.382653         0.166667   

      Physical-Waist_Circumference  Physical-Diastolic_BP  Physical-HeartRate  \
2349                      0.380952               0.423729            0.520548   
344                       0.571429               0.457627            0.287671   
1675                      0.142857               0.423729            0.369863   
1527                      0.190476               0.389831            0.383562   
581                       0.571429               0.423729            0.684932   
...                            ...                    ...                 ...   
2255                      0.190476               0.423729            0.794521   
610                       0.285714               0.000000            0.260274   
1468                      0.095238               0.508475            0.438356   
154                       0.476190               0.423729            0.808219   
2964                      0.190476               0.118644            0.246575   

      ...  PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
2349  ...                  0.0                  0.0              1.0   
344   ...                  1.0                  0.0              0.0   
1675  ...                  0.0                  0.0              1.0   
1527  ...                  0.0                  0.0              1.0   
581   ...                  1.0                  0.0              0.0   
...   ...                  ...                  ...              ...   
2255  ...                  1.0                  0.0              0.0   
610   ...                  0.0                  0.0              0.0   
1468  ...                  0.0                  1.0              0.0   
154   ...                  0.0                  0.0              0.0   
2964  ...                  0.0                  0.0              1.0   

      SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
2349                0.0                0.0                0.0   
344                 1.0                0.0                0.0   
1675                0.0                0.0                0.0   
1527                0.0                0.0                0.0   
581                 0.0                1.0                0.0   
...                 ...                ...                ...   
2255                0.0                1.0                0.0   
610                 1.0                0.0                0.0   
1468                1.0                0.0                0.0   
154                 1.

In [69]:
dfTrain.describe()

Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  Physical-BMI  \
count      2859.000000      2859.000000      2859.000000   2859.000000   
mean          0.328670         0.372858         0.501265      0.447993   
std           0.214577         0.483649         0.194184      0.160103   
min           0.000000         0.000000         0.000000      0.000000   
25%           0.133333         0.000000         0.400000      0.328874   
50%           0.266667         0.000000         0.500000      0.406631   
75%           0.466667         1.000000         0.666667      0.530495   
max           1.000000         1.000000         1.000000      1.000000   

       Physical-Height  Physical-Weight  Physical-Waist_Circumference  \
count      2859.000000      2859.000000                   2859.000000   
mean          0.476068         0.307285                      0.333268   
std           0.169162         0.192733                      0.203139   
min           0.000000         0.000000                      0.000000   
25%           0.350765         0.156566                      0.190476   
50%           0.459184         0.257576                      0.285714   
75%           0.599490         0.434028                      0.428571   
max           1.000000         1.000000                      1.000000   

       Physical-Diastolic_BP  Physical-HeartRate  Physical-Systolic_BP  ...  \
count            2859.000000         2859.000000           2859.000000  ...   
mean                0.485046            0.503927              0.471207  ...   
std                 0.168056            0.180239              0.169735  ...   
min                 0.000000            0.000000              0.000000  ...   
25%                 0.372881            0.383562              0.361111  ...   
50%                 0.474576            0.493151              0.444444  ...   
75%                 0.593220            0.630137              0.569444  ...   
max                 1.000000            1.000000              1.000000  ...   

       PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
count          2859.000000          2859.000000      2859.000000   
mean              0.241343             0.233298         0.238895   
std               0.427973             0.423005         0.426483   
min               0.000000             0.000000         0.000000   
25%               0.000000             0.000000         0.000000   
50%               0.000000             0.000000         0.000000   
75%               0.000000             0.000000         0.000000   
max               1.000000             1.000000         1.000000   

       SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
count        2859.000000        2859.000000        2859.000000   
mean            0.280867           0.237845           0.242392   
std             0.449501           0.425838           0.428605   
min             0.000000           0.000000           0.000000   
25%             0.000000           0.000000           0.000000   
50%             0.000000           0.000000           0.000000   
75%             1.000000           0.000000           0.000000   
max             1.000000           1.000000           1.000000   

       PreInt_EduHx-Season_Fall  PreInt_EduHx-Season_Spring  \
count               2859.000000                 2859.000000   
mean                   0.227702                    0.293809   
std                    0.419423                    0.455585   
min                    0.000000                    0.000000   
25%                    0.000000                    0.000000   
50%                    0.000000                    0.000000   
75%                    0.000000                    1.000000   
max                    1.000000                    1.000000   

       PreInt_EduHx-Season_Summer  PreInt_EduHx-Season_Winter  
count                 2859.000000                 2859.000000  
mean                     0.215110                    0.263379  
st

In [70]:
dfTest.describe()

Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  Physical-BMI  \
count       317.000000       317.000000       317.000000    317.000000   
mean          0.344479         0.410095         0.504259      0.455665   
std           0.212749         0.492628         0.192694      0.160699   
min           0.000000         0.000000         0.000000      0.000000   
25%           0.200000         0.000000         0.416667      0.332789   
50%           0.333333         0.000000         0.500000      0.414396   
75%           0.466667         1.000000         0.666667      0.539056   
max           1.000000         1.000000         0.933333      1.000000   

       Physical-Height  Physical-Weight  Physical-Waist_Circumference  \
count       317.000000       317.000000                    317.000000   
mean          0.477489         0.316923                      0.330607   
std           0.171580         0.188296                      0.200376   
min           0.076531         0.036616                      0.000000   
25%           0.357143         0.181818                      0.190476   
50%           0.459184         0.276515                      0.285714   
75%           0.593112         0.415404                      0.428571   
max           0.918367         0.949495                      0.952381   

       Physical-Diastolic_BP  Physical-HeartRate  Physical-Systolic_BP  ...  \
count             317.000000          317.000000            317.000000  ...   
mean                0.485270            0.496348              0.461006  ...   
std                 0.162055            0.171451              0.173041  ...   
min                 0.000000            0.082192              0.000000  ...   
25%                 0.372881            0.383562              0.333333  ...   
50%                 0.491525            0.493151              0.444444  ...   
75%                 0.593220            0.602740              0.569444  ...   
max                 0.915254            1.000000              0.986111  ...   

       PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
count           317.000000           317.000000       317.000000   
mean              0.236593             0.223975         0.268139   
std               0.425662             0.417564         0.443691   
min               0.000000             0.000000         0.000000   
25%               0.000000             0.000000         0.000000   
50%               0.000000             0.000000         0.000000   
75%               0.000000             0.000000         1.000000   
max               1.000000             1.000000         1.000000   

       SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
count         317.000000         317.000000         317.000000   
mean            0.255521           0.252366           0.223975   
std             0.436843           0.435057           0.417564   
min             0.000000           0.000000           0.000000   
25%             0.000000           0.000000           0.000000   
50%             0.000000           0.000000           0.000000   
75%             1.000000           1.000000           0.000000   
max             1.000000           1.000000           1.000000   

       PreInt_EduHx-Season_Fall  PreInt_EduHx-Season_Spring  \
count                317.000000                  317.000000   
mean                   0.211356                    0.290221   
std                    0.408916                    0.454582   
min                    0.000000                    0.000000   
25%                    0.000000                    0.000000   
50%                    0.000000                    0.000000   
75%                    0.000000                    1.000000   
max                    1.000000                    1.000000   

       PreInt_EduHx-Season_Summer  PreInt_EduHx-Season_Winter  
count                  317.000000                  317.000000  
mean                     0.255521                    0.242902  
st

Export to csvs!

In [70]:
dfTrain.columns[dfTrain.columns.str.contains("PCIAT-PCIAT")]

Index(['PCIAT-PCIAT_01', 'PCIAT-PCIAT_02', 'PCIAT-PCIAT_03', 'PCIAT-PCIAT_04',
       'PCIAT-PCIAT_05', 'PCIAT-PCIAT_06', 'PCIAT-PCIAT_07', 'PCIAT-PCIAT_08',
       'PCIAT-PCIAT_09', 'PCIAT-PCIAT_10', 'PCIAT-PCIAT_11', 'PCIAT-PCIAT_12',
       'PCIAT-PCIAT_13', 'PCIAT-PCIAT_14', 'PCIAT-PCIAT_15', 'PCIAT-PCIAT_16',
       'PCIAT-PCIAT_17', 'PCIAT-PCIAT_18', 'PCIAT-PCIAT_19', 'PCIAT-PCIAT_20',
       'PCIAT-PCIAT_Total'],
      dtype='object')

In [71]:
len(dfTrain.columns[dfTrain.columns.str.contains("PCIAT-PCIAT")])

21

In [72]:
dfTrain.shape

(2859, 107)

In [73]:
dfTrain = dfTrain.drop(columns=dfTrain.columns[dfTrain.columns.str.contains("PCIAT-PCIAT")])
dfTrain.shape

(2859, 86)

In [74]:
dfTest.shape

(317, 107)

In [75]:
dfTest = dfTest.drop(columns=dfTest.columns[dfTest.columns.str.contains("PCIAT-PCIAT")])
dfTest.shape

(317, 86)

In [71]:
dfTrain.to_csv('content/dataTrain.csv', index=False)
dfTest.to_csv('content/dataTest.csv', index=False)